<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Demultiplexing · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="index.html"><img class="logo" src="assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="index.html">Home</a></li><li><a class="toctext" href="symbols.html">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li><a class="toctext" href="sequences/sequences.html">Overview</a></li><li><a class="toctext" href="sequences/bioseq.html">BioSequence</a></li><li><a class="toctext" href="sequences/refseq.html">Reference Sequences</a></li><li><a class="toctext" href="sequences/kmer.html">Nucleic acid k-mers</a></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="io/fasta.html">FASTA formatted files</a></li><li><a class="toctext" href="io/fastq.html">FASTQ formatted files</a></li><li><a class="toctext" href="io/twobit.html">2bit formatted files</a></li></ul></li><li><a class="toctext" href="search.html">Searching</a></li><li><a class="toctext" href="composition.html">Sequence Composition</a></li><li class="current"><a class="toctext" href="demultiplexer.html">Demultiplexing</a><ul class="internal"></ul></li><li><a class="toctext" href="contributing.html">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href="demultiplexer.html">Demultiplexing</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/demultiplexer.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Demultiplexing</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Sequence-demultiplexing-1" href="#Sequence-demultiplexing-1">Sequence demultiplexing</a></h1><p>Multiplex sequencing is a technology to sequence multiple samples at the same time on a high-throughput DNA sequencer. Samples are distinguished by the short prefix of a DNA sequence called DNA barcode. The <code>BioSequences</code> offers the <code>Demultiplexer</code> type and the <code>demultiplex</code> function to identify the DNA barcode of a longer DNA sequence allowing small errors.</p><p>In the following example, four kinds of DNA sequences of length 4 are used as DNA barcodes. <code>Demultiplexer</code> takes these barcodes as its first argument with a few options:</p><pre><code class="language-julia-repl">julia&gt; barcodes = DNASequence[&quot;ATGG&quot;, &quot;CAGA&quot;, &quot;GGAA&quot;, &quot;TACG&quot;];

julia&gt; dplxr = Demultiplexer(barcodes, n_max_errors=1, distance=:hamming)
Demultiplexer{BioSequence{DNAAlphabet{4}}}:
  distance: hamming
  number of barcodes: 4
  number of correctable errors: 1</code></pre><p><code>n_max_errors</code> specifies the number of maximum correctable errors in a barcode. The type of correctable errors depends on the <code>distance</code> parameter. When <code>distance = :hamming</code> as shown above only substitutions are correctable. When <code>distance = :levenshtein</code> substitutions, deletions, and insertions are correctable. The user is responsible for keeping enough distances among barcodes; <code>Demultiplexer</code> will throw an exception if two barcodes are within <code>n_max_errors * 2</code>.</p><p>The <code>demultiplex</code> function takes a demultiplexer object and a DNA sequence, and returns a tuple of a barcode index and a distance between the original barcode sequence and the prefix sequence:</p><pre><code class="language-julia-repl">julia&gt; demultiplex(dplxr, dna&quot;ATGGCGNT&quot;)  # 1st barcode with no errors
(1, 0)

julia&gt; demultiplex(dplxr, dna&quot;CAGGCGNT&quot;)  # 2nd barcode with one error
(2, 1)

julia&gt; demultiplex(dplxr, dna&quot;GGAACGNT&quot;)  # 3rd barcode with no errors
(3, 0)

julia&gt; demultiplex(dplxr, dna&quot;TGACCGNT&quot;)  # no matching barcode
(0, -1)
</code></pre><p>The optional third argument controls the search strategy. <code>demultiplex</code> uses an index to search the closest barcode within <code>n_max_errors</code> in the barcode set and returns it if any by default. If the third argument is <code>true</code> it falls back to a linear search after the index search and returns one of the closest barcodes at random. The next example shows the difference of these two strategies:</p><pre><code class="language-julia-repl">julia&gt; demultiplex(dplxr, dna&quot;TGACCGNT&quot;, false)  # linear search off (default)
(0, -1)

julia&gt; demultiplex(dplxr, dna&quot;TGACCGNT&quot;, true)   # linear search on
(3, 2)
</code></pre><footer><hr/><a class="previous" href="composition.html"><span class="direction">Previous</span><span class="title">Sequence Composition</span></a><a class="next" href="contributing.html"><span class="direction">Next</span><span class="title">Contributing</span></a></footer></article></body></html>
